includes/forms/controllers/class.llms.controller.account.php

```diff
@@ -121,108 +128,149 @@ public function update() {
 		}
 
-		// verify required field
-		if ( empty( $_POST['llms_login'] ) ) {
-			return llms_add_notice( __( 'Enter a username or e-mail address.', 'lifterlms' ), 'error' );
-		}
+		$err = new WP_Error();
+
+		/**
+		 * Fire an action immediately prior to the lost password form submission processing.
+		 *
+		 * @since [version]
+		 */
+		do_action( 'llms_before_lost_password_form_submit' );
 
 		$login = llms_filter_input( INPUT_POST, 'llms_login', FILTER_SANITIZE_STRING );
 
-		// always check email
-		$get_by = array( 'email' );
-		// add login if username generation is disabled (eg users create their own usernames)
-		if ( 'no' === get_option( 'lifterlms_registration_generate_username' ) ) {
-			$get_by[] = 'login';
-		}
+		// Login is required.
+		if ( empty( $login ) ) {
+			$err->add( 'llms_pass_reset_missing_login', __( 'Enter a username or e-mail address.', 'lifterlms' ) );
+		} else {
 
-		$user = null;
-		// check each field to find the user
-		foreach ( $get_by as $field ) {
-			$user = get_user_by( $field, $login );
-			// if we find a user skip the next check
-			if ( $user ) {
-				break;
+			// Locate the user.
+			$field = strpos( $login, '@' ) ? 'email' : 'login';
+			$user  = get_user_by( $field, $login );
+
+			// No user found.
+			if ( ! $user ) {
+
+				$err->add( 'llms_pass_reset_invalid_login', __( 'Invalid username or e-mail address.', 'lifterlms' ) );
+
+			} else {
+
+				/**
+				 * Mimic WordPress core functionality to prevent resetting of password generally or for a specific user.
+				 *
+				 * Returning `false` or a `WP_Error` will prevent password resetting.
+				 *
+				 * If a `WP_Error` is returned, the first error message from the object will be output as an error notice.
+				 *
+				 * @since [version]
+				 *
+				 * @link https://developer.wordpress.org/reference/hooks/allow_password_reset/
+				 *
+				 * @param bool $allow   Whether or not to allow password reset.
+				 * @param int  $user_id WP_User ID of the user who's password is being reset.
+				 */
+				$allow = apply_filters( 'allow_password_reset', true, $user->ID ); // WP core filter.
+
+				if ( ! $allow ) {
+					$err->add( 'llms_pass_reset_disabled', __( 'Password reset is not allowed for this user.', 'lifterlms' ) );
+				} elseif ( is_wp_error( $allow ) ) {
+					$err = $allow;
+				}
 			}
 		}
 
-		// if we don't have a user return an error
-		if ( ! $user ) {
-			return llms_add_notice( __( 'Invalid username or e-mail address.', 'lifterlms' ), 'error' );
+		/**
+		 * Fires before errors are returned from a password reset request.
+		 *
+		 * Mimics WordPress core behavior so 3rd parties don't need to add special handlers for LifterLMS
+		 * password reset flows.
+		 *
+		 * @since [version]
+		 *
+		 * @link https://developer.wordpress.org/reference/hooks/lostpassword_post/
+		 *
+		 * @param WP_Error $err A WP_Error object containing any errors generated by using invalid credentials.
+		 */
+		do_action( 'lostpassword_post', $err );
```

Comment:
At this point, why we don't also pass the 3rd parameter (`WP_User` or `false`), as the original wp-core hook? https://developer.wordpress.org/reference/hooks/lostpassword_post/
In my opinion we always should ensure the original number of parameters, it's up to the listener deciding how many it needs.

Reply:
I believe this should be fired before `allow_password_reset` to match the WordPress core hook order: https://github.com/WordPress/wordpress-develop/blob/master/src/wp-login.php#L388

Reply:
@TimothyBJacobs you mean at line 169 then, makes sense...

Reply:
> At this point, why we don't also pass the 3rd parameter (`WP_User` or `false`), as the original wp-core hook? https://developer.wordpress.org/reference/hooks/lostpassword_post/
> In my opinion we always should ensure the original number of parameters, it's up to the listener deciding how many it needs.

Yea, this isn't an opinion. This is an objective fact. I had the second paraemeter in there and then I removed it right before committing... Why? Because the docs I was looking at are out of date. Why do I try to save time by using stuff like this: https://devdocs.io/wordpress/hooks/lostpassword_post

I'll fix it, thanks